### PR TITLE
Replace some REGEX calls that can be expressed as string functions

### PR DESCRIPTION
**Pull request by JervenBolleman**
Addresses issue: eclipse/rdf4j#1199

This is a trivial query optimizer which can help public endpoints where cheaper string functions (that are easier to further optimize) instead of relatively expensive regex calls. (even if the regex is otherwise trivial in runtime costs).

The regex pattern is tested to make sure it does not contain any of the control characters and is case sensitive. This could be improved to avoid false negatives by inspecting the Pattern object itself. However,
that needs reflection and I think that is not welcome.

This pull request just adds the optimizer, it does not enable it any repository.

There is an copyright assignment from me (and my employer) to Aduna on file. Arranging a new one for RDF4J will take a lot of time, so I hope this acceptable as is.

### CODE REVIEW

**Comment by jeenbroekstra** [comment]
Could you please digitally sign the Eclipse Contributor Agreement (ECA)? See the [Contributor Guidelines](https://github.com/eclipse/rdf4j-doc/blob/master/.github/CONTRIBUTING.md) for more details on how to sign. Unfortunately, we cannot accept your contribution until you have signed this. It shouldn't take more than 5 minutes.

File: evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/RegexAsStringFunctionOptimizer.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
+/** *****************************************************************************
+ * Copyright (c) 2018 Eclipse RDF4J contributors, Aduna, and others.
```
Comment (jeenbroekstra) [inline evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/RegexAsStringFunctionOptimizer.java]: This is the wrong license header. Please use 

```
/*******************************************************************************
 * Copyright (c) 2018 Eclipse RDF4J contributors.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Distribution License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/org/documents/edl-v10.php.
 *******************************************************************************/
```

File: evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/RegexAsStringFunctionOptimizierTest.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
+/** *****************************************************************************
+ * Copyright (c) 2018 Eclipse RDF4J contributors, Aduna, and others.
```
Comment (jeenbroekstra) [inline evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/RegexAsStringFunctionOptimizierTest.java]: Ditto here.

**Comment by jeenbroekstra** [comment]
We'll need to rebase this PR to the develop branch before merging (I tried just now but it messes up the diff a bit, so decided to undo until we're good to go with the merge)

**Comment by JervenBolleman** [comment]
I am struggling with the eclipse account signup :( but will do so.

**Comment by jeenbroekstra** [comment]
@JervenBolleman I've noticed the Eclipse site was a bit slow and flaky today. Let me know if you hit a wall and I'll try and get a maintainer involved.

**Comment by JervenBolleman** [comment]
ECA is signed, contribution SOP inside my workplace followed and not objected to so allowed to make the contribution.

Linking eclipse account to github id does not seem to work nicely so you might need to do some manual work to link it.

**Comment by JervenBolleman** [comment]
The e-mail is jerven.bolleman@sib.swiss

**Comment by jeenbroekstra** [comment]
Thanks, confirmed!

File: evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/RegexAsStringFunctionOptimizierTest.java (hunk 2 above)
Comment (jeenbroekstra) [reply evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/RegexAsStringFunctionOptimizierTest.java]:
```suggestion
 * Copyright (c) 2018 Eclipse RDF4J contributors.
```

**Comment by JervenBolleman** [comment]
Sorry for missing the second license header.
Is the content of the patch itself ok? 
No regexes that could be converted when they shouldn't be?

**Comment by jeenbroekstra** [comment]
Thanks for fixing @JervenBolleman . I would normally just correct such small things myself but for some reason your PR is not editable for me.

 Anyway: content looks ok to me, you've got a fair spread of test cases. A more comprehensive test would be to run a full SPARQL WG testsuite with this optimizer enabled, and see what happens. In fact it might be a good idea to do that as part of this PR, if you're up for it. 

We should also have a think about whether we want this enabled in the core stores (memory and native).

**Comment by jeenbroekstra** [comment]
@rdf4j-bot run tests

**Comment by JervenBolleman** [comment]
I am ok to run the full test suite. I think it makes sense to enable it for the native and memory stores. The evaluation time to determine if this optimizer should run is minimal and not enabled code tends to bitrot.
While gains are minimal so are costs.

In any case I started with this minimal optimizer before starting on more complicated ones to get familiar with the process.

**Comment by jeenbroekstra** [comment]
@JervenBolleman are you able to update this PR to enable this optimizer in the Memory- and NativeStore by default? That should also automatically tell us if we've overlooked any corner cases, as it will then be used in all our compliance tests.

**Comment by JervenBolleman** [comment]
@rdf4j-bot run tests

@jeenbroekstra sorry between travel and christmas I misunderstood what you liked. I think the last commit should turn everything on.

**Comment by JervenBolleman** [comment]
Sorry for force pushing but that fixes the ECA check...

**Comment by jeenbroekstra** [comment]
@rdf4j-bot run tests
